File: fbo-launchpad-backend/src/models/fuel_order.py
import enum
from datetime import datetime
from sqlalchemy import Integer, String, Boolean, DateTime, Enum, Text, Numeric, ForeignKey
from src.app import db

class FuelOrderStatus(enum.Enum):
    DISPATCHED = 'Dispatched'
    ACKNOWLEDGED = 'Acknowledged'
    EN_ROUTE = 'En Route'
    FUELING = 'Fueling'
    COMPLETED = 'Completed'
    REVIEWED = 'Reviewed'
    CANCELLED = 'Cancelled'

class FuelOrder(db.Model):
    __tablename__ = 'fuel_orders'

    # Primary Key
    id = db.Column(db.Integer, primary_key=True)

    # Status and Core Fields
    status = db.Column(db.Enum(FuelOrderStatus), nullable=False, default=FuelOrderStatus.DISPATCHED, index=True)
    tail_number = db.Column(db.String(20), db.ForeignKey('aircraft.tail_number'), nullable=False, index=True)
    customer_id = db.Column(db.Integer, db.ForeignKey('customers.id'), nullable=True)
    fuel_type = db.Column(db.String(50), nullable=False)
    additive_requested = db.Column(db.Boolean, default=False)
    requested_amount = db.Column(db.Numeric(10, 2), nullable=True)

    # Assignment Fields
    assigned_lst_user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=True, index=True)
    assigned_truck_id = db.Column(db.Integer, db.ForeignKey('fuel_trucks.id'), nullable=True, index=True)
    location_on_ramp = db.Column(db.String(100), nullable=True)
    
    # Notes Fields
    csr_notes = db.Column(db.Text, nullable=True)
    lst_notes = db.Column(db.Text, nullable=True)

    # Metering Fields
    start_meter_reading = db.Column(db.Numeric(12, 2), nullable=True)
    end_meter_reading = db.Column(db.Numeric(12, 2), nullable=True)
    calculated_gallons_dispensed = db.Column(db.Numeric(10, 2), nullable=True)

    # Timestamps
    created_at = db.Column(db.DateTime, nullable=False, default=datetime.utcnow)
    dispatch_timestamp = db.Column(db.DateTime, nullable=True)
    acknowledge_timestamp = db.Column(db.DateTime, nullable=True)
    en_route_timestamp = db.Column(db.DateTime, nullable=True)
    fueling_start_timestamp = db.Column(db.DateTime, nullable=True)
    completion_timestamp = db.Column(db.DateTime, nullable=True)
    reviewed_timestamp = db.Column(db.DateTime, nullable=True)
    
    # Review Fields
    reviewed_by_csr_user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=True)

    # Relationships
    aircraft = db.relationship('Aircraft', backref=db.backref('fuel_orders', lazy='dynamic'))
    customer = db.relationship('Customer', backref=db.backref('fuel_orders', lazy='dynamic'))
    assigned_lst = db.relationship('User', foreign_keys=[assigned_lst_user_id], 
                                 backref=db.backref('assigned_fuel_orders', lazy='dynamic'))
    assigned_truck = db.relationship('FuelTruck', backref=db.backref('fuel_orders', lazy='dynamic'))
    reviewed_by_csr = db.relationship('User', foreign_keys=[reviewed_by_csr_user_id], 
                                    backref=db.backref('reviewed_fuel_orders', lazy='dynamic'))

    def __repr__(self):
        return f'<FuelOrder {self.id} - {self.tail_number}>' 

File: fbo-launchpad-frontend-csr/src/types/fuelOrder.ts
export enum FuelOrderStatus {
  PENDING = 'PENDING',
  DISPATCHED = 'DISPATCHED',
  ACKNOWLEDGED = 'ACKNOWLEDGED',
  EN_ROUTE = 'EN_ROUTE',
  FUELING = 'FUELING',
  COMPLETED = 'COMPLETED',
  REVIEWED = 'REVIEWED',
  CANCELLED = 'CANCELLED',
  ON_HOLD = 'ON_HOLD' // Added based on common FBO scenarios, confirm if used
}

export interface FuelOrder {
  id: number;
  tail_number: string;
  customer_id?: number | null;
  fuel_type: string;
  additive_requested: boolean;
  requested_amount: string | null; // Backend returns as string (Decimal)
  assigned_lst_user_id: number | null;
  assigned_truck_id: number | null;
  location_on_ramp: string;
  csr_notes?: string | null;
  status: FuelOrderStatus | string;
  created_at: string; // ISO date string
  updated_at?: string; // ISO date string
  start_meter_reading?: string | null; // Decimal
  end_meter_reading?: string | null; // Decimal
  calculated_gallons_dispensed?: string | null; // Decimal
  lst_notes?: string | null;
  dispatch_timestamp?: string | null;
  acknowledge_timestamp?: string | null;
  en_route_timestamp?: string | null;
  fueling_start_timestamp?: string | null;
  completion_timestamp?: string | null;
  reviewed_timestamp?: string | null;
  reviewed_by_csr_user_id?: number | null;
  // Add other fields from detailed view if needed (e.g., aircraft, customer details if joined)
}

export interface PaginatedFuelOrdersResponse {
  orders: FuelOrder[];
  message: string;
  pagination: {
    page: number;
    per_page: number;
    total: number;
    pages: number;
    has_next: boolean;
    has_prev: boolean;
  };
}

export interface FuelOrderStatusCounts {
  [key: string]: number; // e.g., { "PENDING": 5, "COMPLETED": 10 }
}

export interface StatusCountsResponse {
    message: string;
    counts: FuelOrderStatusCounts;
}

// Add other Fuel Order related types here as the UI develops, for example:
// export interface FuelOrderFilters {
//   status?: FuelOrderStatus | string;
//   tail_number?: string;
//   assigned_lst_user_id?: number;
//   assigned_truck_id?: number;
//   date_from?: string;
//   date_to?: string;
// } 

// --- Interfaces for Request Payloads (from FuelOrderService.ts) ---

export interface CreateFuelOrderPayload {
  tail_number: string;
  fuel_type: string;
  assigned_lst_user_id: number; // -1 for auto-assign
  assigned_truck_id: number; // -1 for auto-assign
  requested_amount: number;
  location_on_ramp: string;
  aircraft_type?: string;
  customer_id?: number;
  additive_requested?: boolean;
  csr_notes?: string;
}

export interface UpdateFuelOrderStatusPayload {
  status: FuelOrderStatus | string; // Allow string for flexibility, backend expects uppercase enum string
  assigned_truck_id: number;
}

export interface SubmitFuelDataPayload {
  start_meter_reading: number;
  end_meter_reading: number;
  lst_notes?: string;
}

export interface UpdateFuelOrderPayload {
  tail_number?: string;
  fuel_type?: string;
  assigned_lst_user_id?: number;
  assigned_truck_id?: number;
  requested_amount?: number;
  location_on_ramp?: string;
  aircraft_type?: string; // Or aircraft_id if that's what your system uses
  customer_id?: number;
  additive_requested?: boolean;
  csr_notes?: string;
  // Fields like 'status' are typically updated via specific status change endpoints,
  // so it might be excluded here unless general edit can also change it.
  // For now, mirroring CreateFuelOrderPayload closely, excluding status.
  priority?: 'Low' | 'Normal' | 'High'; // If priority is stored and editable
} 